src/usexp/lexer.mll

```diff
@@ -116,38 +227,51 @@ and jbuild_atom acc start = parse
     { lexbuf.lex_start_p <- start;
       error lexbuf "jbuild_atoms cannot contain |#"
     }
+  | "${" {
+      jbuild_atom ((template_variable Dollar_brace lexbuf) :: acc) start lexbuf }
   | ('#'+ | '|'+ | (atom_char # ['|' '#'])) as s
-    { jbuild_atom (if acc = "" then s else acc ^ s) start lexbuf
-    }
-  | ""
-    { if acc = "" then
-        error lexbuf "Internal error in the S-expression parser, \
-                      please report upstream.";
-      lexbuf.lex_start_p <- start;
-      Token.Atom (A acc)
-    }
+    { jbuild_atom (Template.add_text acc s) start lexbuf }
+  | "" { Template.token acc ~quoted:false ~start lexbuf }
 
 and quoted_string mode = parse
   | '"'
-    { Buffer.contents escaped_buf }
+    { Template.Buffer.get () }
   | '\\'
     { match escape_sequence mode lexbuf with
       | Newline -> quoted_string_after_escaped_newline mode lexbuf
       | Other   -> quoted_string                       mode lexbuf
     }
+  | (['$' '%'] as varchar) (['{' '('] as delim) {
```

Comment:
Is `%(` intended to be a valid combination here, or is this filtered in another place?